[scroll-animations] https://scroll-driven-animations.style/demos/parallax-carousel/css/ does not work

graouts opened the pull request:
#### 319e2fc273efed67ae7e06d3daf850fbaef86e5d
<pre>
[scroll-animations] <a href="https://scroll-driven-animations.style/demos/parallax-carousel/css/">https://scroll-driven-animations.style/demos/parallax-carousel/css/</a> does not work
<a href="https://bugs.webkit.org/show_bug.cgi?id=284534">https://bugs.webkit.org/show_bug.cgi?id=284534</a>
<a href="https://rdar.apple.com/141356306">rdar://141356306</a>

Reviewed by Antti Koivisto.

The Scroll-driven Animations demo at <a href="https://scroll-driven-animations.style/demos/parallax-carousel/css/">https://scroll-driven-animations.style/demos/parallax-carousel/css/</a> uses
an `animation-timeline: view()` property set via a CSS rule on a number of sibling `&lt;img&gt;` elements. In WebKit,
these elements share style.

However, the anonymous `ViewTimeline` created by `animation-timeline: view()` is stored on an `Animation` object
owned by `RenderStyle` through its `AnimationList` for CSS Animations accessed via `RenderStyle::animations()`.
If elements that are using an anonymous timeline share style, then they&apos;ll share their timeline even though those
need to be unique to each element.

To address this, we no longer create a `ScrollTimeline` or `ViewTimeline` when finding an anonymous timeline set
for the `animation-timeline` property and instead use a dedicated small struct for each type which we add to the
`std::variant&lt;&gt;` used to represent timelines on `Animation`. Then, when we deal with a style change on the associated
`CSSAnimation`, we create a new `ScrollTimeline` or `ViewTimeline` using the struct hosted on `Animation`.

This change yields a fair bit of refactoring to deal with the expanded variant but yields some WPT progressions as
well as fixing the demo.

We also add a new WPT test specifically testing timeline uniqueness for this case.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-nearest-dirty-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-root-dirty-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/contain-alignment-expected.txt:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::create):
(WebCore::operator&lt;&lt;):
(WebCore::ScrollTimeline::createFromCSSValue): Deleted.
(WebCore::ScrollTimeline::dump const): Deleted.
(WebCore::ScrollTimeline::toCSSValue const): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::createFromCSSValue): Deleted.
(WebCore::ViewTimeline::dump const): Deleted.
(WebCore::ViewTimeline::toCSSValue const): Deleted.
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/css/CSSScrollValue.h:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapAnimationTimeline const):
* Source/WebCore/css/CSSViewValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationTimeline):
* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::AnonymousScrollTimeline::operator== const):
(WebCore::Animation::AnonymousViewTimeline::operator== const):

Canonical link: <a href="https://commits.webkit.org/287836@main">https://commits.webkit.org/287836@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0116a161e54f8cbdc0509caeab697571f15a9730

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81050 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85579 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32036 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8377 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63267 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/21032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84119 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/353 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73767 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43565 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/250 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27928 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30494 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/71795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28482 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87014 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8280 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/5844 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71570 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8457 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70806 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14851 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/13775 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12567 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8241 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/13764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8078 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11598 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9886 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk reviewed (approved):
much better!

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1613,13 +1613,46 @@ static Ref<CSSPrimitiveValue> valueForScopedName(const Style::ScopedName& scoped
+                return CSSValueNearest;
+            }
+        }();
+        return CSSScrollValue::create(CSSPrimitiveValue::create(scroller), CSSPrimitiveValue::create(toCSSValueID(anonymousScrollTimeline.axis)));
```
```suggestion
        return CSSScrollValue::create(CSSPrimitiveValue::create(scroller), createConvertingToCSSValueID(anonymousScrollTimeline.axis));
```

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1613,13 +1613,46 @@ static Ref<CSSPrimitiveValue> valueForScopedName(const Style::ScopedName& scoped
+        };
+
+        return CSSViewValue::create(
+            CSSPrimitiveValue::create(toCSSValueID(anonymousViewTimeline.axis)),
```
```suggestion
            createConvertingToCSSValueID(anonymousViewTimeline.axis),
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html:
```diff
@@ -0,0 +1,78 @@
+<script src="/resources/testharnessreport.js"></script>
+<script src="/scroll-animations/scroll-timelines/testcommon.js"></script>
+<title>The "animation-timeline" CSS property yields a unique anonymous timeline for siblings matching the same CSS rules</title>
+</head>
```
`</head>` should close after style.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html:
```diff
@@ -0,0 +1,78 @@
+  </div>
+
+</body>
+<script>
```
script should be inside body

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html:
```diff
@@ -0,0 +1,78 @@
+  }, `Setting "animation-timeline: ${type}()" yields a unique ${type} timeline for siblings matching the same CSS rules`);
+});
+
+</script>
```
missing `</html>`

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html:
```diff
@@ -0,0 +1,78 @@
+<meta charset="utf-8">
+<link rel="help" src="https://drafts.csswg.org/scroll-animations-1/#scroll-timelines-anonymous">
+<link rel="help" src="https://drafts.csswg.org/scroll-animations-1/#view-timelines-anonymous">
+<script src="/resources/testharness.js"></script>
+<script src="/resources/testharnessreport.js"></script>
+<script src="/scroll-animations/scroll-timelines/testcommon.js"></script>
```
I'd move this down into the body before the main script

nt1m reviewed (commented):
some nits on test formatting

graouts committed:
[scroll-animations] https://scroll-driven-animations.style/demos/parallax-carousel/css/ does not work
https://bugs.webkit.org/show_bug.cgi?id=284534
rdar://141356306

Reviewed by Antti Koivisto.

The Scroll-driven Animations demo at https://scroll-driven-animations.style/demos/parallax-carousel/css/ uses
an `animation-timeline: view()` property set via a CSS rule on a number of sibling `<img>` elements. In WebKit,
these elements share style.

However, the anonymous `ViewTimeline` created by `animation-timeline: view()` is stored on an `Animation` object
owned by `RenderStyle` through its `AnimationList` for CSS Animations accessed via `RenderStyle::animations()`.
If elements that are using an anonymous timeline share style, then they'll share their timeline even though those
need to be unique to each element.

To address this, we no longer create a `ScrollTimeline` or `ViewTimeline` when finding an anonymous timeline set
for the `animation-timeline` property and instead use a dedicated small struct for each type which we add to the
`std::variant<>` used to represent timelines on `Animation`. Then, when we deal with a style change on the associated
`CSSAnimation`, we create a new `ScrollTimeline` or `ViewTimeline` using the struct hosted on `Animation`.

This change yields a fair bit of refactoring to deal with the expanded variant but yields some WPT progressions as
well as fixing the demo.

We also add a new WPT test specifically testing timeline uniqueness for this case.

* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-nearest-dirty-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/scroll-timeline-root-dirty-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/css/siblings-with-anonymous-timelines.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-animations/view-timelines/contain-alignment-expected.txt:
* Source/WebCore/animation/CSSAnimation.cpp:
(WebCore::CSSAnimation::syncPropertiesWithBackingAnimation):
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::create):
(WebCore::operator<<):
(WebCore::ScrollTimeline::createFromCSSValue): Deleted.
(WebCore::ScrollTimeline::dump const): Deleted.
(WebCore::ScrollTimeline::toCSSValue const): Deleted.
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::createFromCSSValue): Deleted.
(WebCore::ViewTimeline::dump const): Deleted.
(WebCore::ViewTimeline::toCSSValue const): Deleted.
* Source/WebCore/animation/ViewTimeline.h:
* Source/WebCore/css/CSSScrollValue.h:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapAnimationTimeline const):
* Source/WebCore/css/CSSViewValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForAnimationTimeline):
* Source/WebCore/platform/animation/Animation.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/animation/Animation.h:
(WebCore::Animation::AnonymousScrollTimeline::operator== const):
(WebCore::Animation::AnonymousViewTimeline::operator== const):

Canonical link: https://commits.webkit.org/287836@main
